### ISSUE TITLE
Making API calls on behalf of different senders

### ISSUE THREAD

**Issue by gmile** [EthWorks/ethereum.rb#59]
**TL;DR**. Would `ethereum.rb` team consider to accept a change to make it possible to pass (override) sender during a transaction?

---

Right now it looks like the `ethereum.rb` was designed to make API calls from a certain single address aka `@sender` aka `default_address`. I suspect this was done to support a case, where the library user only has one address, and would like to issue all transactions using that address.

Once initialized, there seems to be no way to override sender on demand. Hence the singleton. In the case of our application, we need to use different senders all the time: we operate multiple addresses and often need to send tokens from these addresses to contract. However looks like is it not possible to do, by design.

To work around this, I have to "manually" write something like this:

```ruby
require 'ethereum'

account_a_address = "0xd8e05701eFf33acfDA0A8e24C04070347703c72C"
account_b_address = "0xad9c9b5085bec8e8013296dc12bd9cecad102356"
contract_address  = "0x5928ead021bd523e6d3531de920d78b7a02efa60"

client = Ethereum::IpcClient.new
encoder = Ethereum::Encoder.new
contract = Ethereum::Contract.create(file: "/path/to/contract_text.sol", address: contract_address)

fun = Ethereum::Abi.parse_abi(contract.abi)[1].find { |fun| fun.name == "transferFrom" }

args = [
  account_a_address,
  account_b_address,
  value_to_transfer = Time.now.to_i
]

contract.client.eth_send_transaction({
  to: contract_address,
  from: account_a_address,
  data: ("0x" + fun.signature + (encoder.encode_arguments(fun.inputs, args).presence || "0"*64)),
  gas_limit: client.int_to_hex(client.gas_limit),
  gas_price: client.int_to_hex(client.gas_price)
})["result"]
```

**Comment by ghivert** [comment]
I got the same problem. Working with directly with Infura, having to change the sender, as no default client was running on my computer.

You can do something like this:

```ruby
contract.sender = address
```

It works for me.

**Comment by gmile** [comment]
@ghivert you're totally right, thank you! Indeed, I figured this out later after posting the issue, but forgot to close it.

**Comment by ToJen** [comment]
@ghivert I'm also using Infura like this:

```ruby
    @client = Ethereum::HttpClient.new('https://ropsten.infura.io/my_api_key')
    PATH = "#{Dir.pwd}/contracts/MyContract.sol"

    def create
        @contract = Ethereum::Contract.create(file: PATH, address: "my_contract_address", client:@client)
        # @contract.key = "an_account_i_created_using_parity_on_ropsten"
        @contract.deploy_and_wait
        ## render html: address ##
        render json: @contract.abi #json: client.eth_gas_price
    end
```

How do I set the default account?

**Comment by jasonadkison** [comment]
@ToJen you set the default account address on the client before sending a transaction:

`@client.default_account = '0x123..'`

You also need to setup the key on the contract:
`@contract.key = Eth::Key.new(priv: your_private_key)`
